File: googology/notations/ordinal/sequence/marked_parent/pps.py
"""
PPS (Parented Predecessor Sequence) — 标记父项型 (marked_parent) 序数序列记号。

由 318`4 创造。PPS 系列 = PPS1（原始）+ PPS2 + PPS4 + 变体 wPPS4 / tPPS4 /
fPPS4；PPS3 / ePPS4 / sPPS4 资料缺失（N/A），不实现。已知结论（外部）：
良序极限 = ζ₀；PPS2 为失败的弱化修复；PPS4 旨在删除无穷降链。本库一律
不求值，expand 只产出一步基本列。

公共骨架（全部变体一致）：
  序列 A = (a_1, ..., a_n)，首项是第 1 项；极限表达式 0,1,2,3,4,5,...
  末项值 x = a_n，末项列标 y = n，坏根 = 第 x 项，坏根值 b = a_x，L = y - x。
  1. 末项为 0 → 后继序数（去掉末项）。
  2. 按变体替换末项（见 _replace_last）。
  3. 复制：对任意 i > x，第 i+L 项由第 i 项确定：a_i >= x → a_i + L，否则 a_i。
  基本列[n] = 展开到第 y + nL - 1 项（n 从 1 起）。库 0 索引：expand(k) =
  基本列[k+1] = expandLen((k+1)L - 1)；expand_to(M) = expandLen(A, M)。

变体差异（仅末项替换 / 强展开复制）：
  PPS1  存在值 = b 的项 → 换为 b；否则减 1。
  PPS2  存在值 = b 的项（取最右，列标 k）时，比较 a_{k+j} 与 a_{c+j}
        （c = x），首个不同处前者更大才换 b，否则减 1。
  PPS4  存在值 = b → 弱展开（换 b）；否则强展开：第 b 列与第 x 列之间
        （不含）最右侧值 = b 的项，末项换为该列标；找不到 → 等同弱展开。
  wPPS4 注记称把 PPS4 强展开判定由 ≤ b 收紧为 = b（PPS4 正文即 = b，
        按字面实现后二者一致；差异来源见 C++ spec）。
  tPPS4 wPPS4 + 强展开时由末项复制出的项加阶差 L。
  fPPS4 tPPS4 + 强展开时末项第一次复制为坏根列标 x。
"""

from enum import Enum

from googology.core import Op, Capabilities, Family, Notation, NotComparable


class PpsVariant(Enum):
    """PPS 变体枚举。"""
    PPS1 = "pps1"
    PPS2 = "pps2"
    PPS4 = "pps4"
    WPPS4 = "wpps4"
    TPPS4 = "tpps4"
    FPPS4 = "fpps4"

    @property
    def version(self) -> str:
        return {"pps1": "1", "pps2": "2", "pps4": "4",
                "wpps4": "4w", "tpps4": "4t", "fpps4": "4f"}[self.value]


class Pps(Notation):
    """PPS 基类：variant + 公共展开骨架。各变体子类仅设定 _variant。"""

    _variant = PpsVariant.PPS1

    def __init__(self, s: str = ""):
        self._seq: list[int] = []
        self._is_master_limit = False
        if s:
            self.string_to_it(s)

    # --- taxonomy ---
    @property
    def name(self) -> str:
        return self._variant.value

    @property
    def family(self) -> Family:
        return Family.Ordinal

    @property
    def subfamily(self) -> str:
        return "sequence"

    @property
    def style(self) -> str:
        return "marked_parent"

    @property
    def creator(self) -> str:
        return "318`4"

    @property
    def version(self) -> str:
        return self._variant.version

    @property
    def capabilities(self) -> Capabilities:
        c = Capabilities()
        for op in (Op.FromString, Op.ToString, Op.Normalize,
                   Op.Compare, Op.Expand, Op.ExpandTo, Op.Successor):
            c.set(op)
        return c

    # --- 辅助 ---
    def _rightmost_eq(self, lo: int, hi: int, b: int) -> int:
        """开区间 (lo, hi) 列标内最右侧值 = b 的列标；无则 0。"""
        for j in range(hi - 1, lo, -1):
            if 1 <= j <= len(self._seq) and self._seq[j - 1] == b:
                return j
        return 0

    def _tail_cmp(self, k: int, c: int) -> int:
        """字典序比较 a_{k+j} 与 a_{c+j}（j = 0,1,2,...）。"""
        y = len(self._seq)
        j = 0
        while True:
            ik, ic = k + j, c + j
            ok_k = 1 <= ik <= y
            ok_c = 1 <= ic <= y
            if not ok_k and not ok_c:
                return 0
            if not ok_k:
                return -1
            if not ok_c:
                return 1
            if self._seq[ik - 1] != self._seq[ic - 1]:
                return 1 if self._seq[ik - 1] > self._seq[ic - 1] else -1
            j += 1

    def _replace_last(self, x: int, b: int, y: int) -> tuple:
        """按变体替换末项，返回 (新末项值, strong)。"""
        v = self._variant
        if v is PpsVariant.PPS1:
            k = self._rightmost_eq(x, y, b)
            return (b if k != 0 else x - 1), False
        if v is PpsVariant.PPS2:
            k = self._rightmost_eq(x, y, b)
            if k != 0 and self._tail_cmp(k, x) > 0:
                return b, False
            return x - 1, False
        # PPS4 家族：弱/强展开二分
        k = self._rightmost_eq(x, y, b)
        if k != 0:
            return b, False
        j = self._rightmost_eq(b, x, b)
        return (j if j != 0 else b), True

    def _copied(self, src: int, x: int, L: int, y: int, i: int, strong: bool) -> int:
        """复制规则：源位置 i（1-based）处值 src 的映射。"""
        base = src + L if src >= x else src
        if not strong:
            return base
        if self._variant in (PpsVariant.TPPS4, PpsVariant.FPPS4):
            if i % L == y % L:  # 末项链（源 i = y, y+L, y+2L, ...）
                if self._variant is PpsVariant.FPPS4 and i == y:
                    return x
                return src + L
        return base

    def _expand_len(self, M: int) -> None:
        """expandLen(A, M)：替换末项 + 追加 M 项（源 i = p + y - L）。"""
        if not self._seq:
            return
        y = len(self._seq)
        x = self._seq[-1]
        if x == 0:
            self._seq.pop()
            return
        if x >= y:
            raise ValueError("Pps: 坏根 = 第 x 项不存在（x >= y）")
        b = self._seq[x - 1]
        L = y - x
        new_last, strong = self._replace_last(x, b, y)
        self._seq[y - 1] = new_last
        for p in range(1, M + 1):
            i = p + y - L
            if i < 1 or i > len(self._seq):
                raise ValueError("Pps: 源索引越界")
            self._seq.append(self._copied(self._seq[i - 1], x, L, y, i, strong))

    # --- string ---
    def string_to_it(self, s: str) -> None:
        t = "".join(ch for ch in s if not ch.isspace())
        while t and t[0] in "([{":
            t = t[1:]
        while t and t[-1] in ")]}":
            t = t[:-1]
        self._seq = [int(p) for p in t.split(",") if p] if t else []

    def to_latex(self) -> str:
        if self._is_master_limit:
            return "(0, 1, 2, …)"
        return "(" + ", ".join(str(v) for v in self._seq) + ")"

    # --- operations ---
    def expand(self, n: int = 1) -> None:
        """库 0 索引：expand(k) = 基本列[k+1] = expandLen((k+1)L - 1)。"""
        if self._is_master_limit:
            self._seq = list(range(n))
            self._is_master_limit = False
            return
        if not self._seq:
            return
        x = self._seq[-1]
        if x == 0:
            self._seq.pop()
            return
        y = len(self._seq)
        if x >= y:
            raise ValueError("Pps: 坏根 = 第 x 项不存在（x >= y）")
        L = y - x
        self._expand_len((n + 1) * L - 1)

    def expand_to(self, length: int) -> None:
        self._expand_len(length)

    def __getitem__(self, n: int) -> "Pps":
        clone = self._clone()
        clone.expand(n)
        return clone

    def _clone(self) -> "Pps":
        c = self.__class__()
        c._seq = list(self._seq)
        c._is_master_limit = self._is_master_limit
        return c

    def compare(self, other: Notation) -> int:
        if not isinstance(other, Pps):
            raise NotComparable(self.name)
        if type(self) is not type(other):
            raise NotComparable(self.name)
        if self._is_master_limit or other._is_master_limit:
            if self._is_master_limit and other._is_master_limit:
                return 0
            return 1 if self._is_master_limit else -1
        a, b = self._seq, other._seq
        for x, y in zip(a, b):
            if x > y:
                return 1
            if x < y:
                return -1
        if len(a) > len(b):
            return 1
        if len(a) < len(b):
            return -1
        return 0

    @property
    def is_successor(self) -> bool:
        return (not self._is_master_limit) and bool(self._seq) and self._seq[-1] == 0

    # --- §12 limit API ---
    @classmethod
    def limit(cls, n: int) -> "Pps":
        p = cls()
        p._seq = list(range(n))
        return p

    @classmethod
    def master_limit(cls) -> "Pps":
        p = cls()
        p._is_master_limit = True
        return p


class Pps1(Pps):
    _variant = PpsVariant.PPS1


class Pps2(Pps):
    _variant = PpsVariant.PPS2


class Pps4(Pps):
    _variant = PpsVariant.PPS4


class WPps4(Pps):
    _variant = PpsVariant.WPPS4


class TPps4(Pps):
    _variant = PpsVariant.TPPS4


class FPps4(Pps):
    _variant = PpsVariant.FPPS4

File: src/notations/ordinal/sequence/marked_parent/pps.rs
//! PPS (Parented Predecessor Sequence) — marked-parent ordinal sequence
//! notation. Created by 318`4. Variants: PPS1 (original) / PPS2 / PPS4 /
//! wPPS4 / tPPS4 / fPPS4. PPS3 / ePPS4 / sPPS4 are N/A (no data) — not
//! implemented.
//!
//! Bad root = the x-th term (x = last value); b = bad-root value; L = y - x
//! (y = last column). expand_len replaces the last term (per variant) and
//! appends elements from the running sequence: a_i >= x -> a_i + L, else a_i.
//! 基本列[n] = expand to the (y + nL - 1)-th term (n >= 1). Library 0-index:
//! expand(k) = 基本列[k+1] = expand_len((k+1)L - 1). compare is lexicographic;
//! cross-variant returns Err(NotComparable).

use std::any::Any;

use crate::core::capability::{Capabilities, Op};
use crate::core::notation::{Family, Notation, NotComparable};

/// PPS variants.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PpsVariant {
    Pps1,
    Pps2,
    Pps4,
    WPps4,
    TPps4,
    FPps4,
}

impl PpsVariant {
    fn reg_name(self) -> &'static str {
        match self {
            PpsVariant::Pps1 => "pps1",
            PpsVariant::Pps2 => "pps2",
            PpsVariant::Pps4 => "pps4",
            PpsVariant::WPps4 => "wpps4",
            PpsVariant::TPps4 => "tpps4",
            PpsVariant::FPps4 => "fpps4",
        }
    }

    fn ver(self) -> &'static str {
        match self {
            PpsVariant::Pps1 => "1",
            PpsVariant::Pps2 => "2",
            PpsVariant::Pps4 => "4",
            PpsVariant::WPps4 => "4w",
            PpsVariant::TPps4 => "4t",
            PpsVariant::FPps4 => "4f",
        }
    }
}

/// PPS state: variant + sequence terms + master-limit flag.
#[derive(Clone, Debug)]
pub struct Pps {
    variant: PpsVariant,
    seq: Vec<i64>,
    is_master_limit: bool,
}

impl Pps {
    pub fn new(v: PpsVariant) -> Self {
        Pps {
            variant: v,
            seq: Vec::new(),
            is_master_limit: false,
        }
    }

    pub fn from_string(v: PpsVariant, s: &str) -> Result<Self, String> {
        let mut p = Pps::new(v);
        p.string_to_it(s)?;
        Ok(p)
    }

    /// limit(n) = (0,1,...,n-1) of a variant.
    pub fn limit(v: PpsVariant, n: i64) -> Self {
        let mut p = Pps::new(v);
        p.seq = (0..n).collect();
        p
    }

    /// Master limit expression of a variant (supremum).
    pub fn master_limit(v: PpsVariant) -> Self {
        let mut p = Pps::new(v);
        p.is_master_limit = true;
        p
    }

    /// Sequence ends with 0 => successor.
    pub fn is_successor(&self) -> bool {
        !self.is_master_limit && self.seq.last() == Some(&0)
    }

    pub fn to_latex(&self) -> String {
        if self.is_master_limit {
            return "(0, 1, 2, …)".to_string();
        }
        let parts: Vec<String> = self.seq.iter().map(|v| v.to_string()).collect();
        format!("({})", parts.join(", "))
    }

    /// Rightmost 1-based column j in the open interval (lo, hi) with value = b.
    fn rightmost_eq(&self, lo: i64, hi: i64, b: i64) -> i64 {
        let mut j = hi - 1;
        while j > lo {
            if j >= 1 && (j as usize) <= self.seq.len() && self.seq[(j - 1) as usize] == b {
                return j;
            }
            j -= 1;
        }
        0
    }

    /// Lexicographic compare of a_{k+j} vs a_{c+j} (j = 0,1,2,...).
    fn tail_cmp(&self, k: i64, c: i64) -> i32 {
        let y = self.seq.len() as i64;
        let mut j = 0i64;
        loop {
            let ik = k + j;
            let ic = c + j;
            let ok_k = ik >= 1 && ik <= y;
            let ok_c = ic >= 1 && ic <= y;
            if !ok_k && !ok_c {
                return 0;
            }
            if !ok_k {
                return -1;
            }
            if !ok_c {
                return 1;
            }
            let vk = self.seq[(ik - 1) as usize];
            let vc = self.seq[(ic - 1) as usize];
            if vk != vc {
                return if vk > vc { 1 } else { -1 };
            }
            j += 1;
        }
    }

    /// Per-variant last-term replacement: (new_last, strong).
    fn replace_last(&self, x: i64, b: i64, y: i64) -> (i64, bool) {
        match self.variant {
            PpsVariant::Pps1 => {
                let k = self.rightmost_eq(x, y, b);
                (if k != 0 { b } else { x - 1 }, false)
            }
            PpsVariant::Pps2 => {
                let k = self.rightmost_eq(x, y, b);
                (if k != 0 && self.tail_cmp(k, x) > 0 { b } else { x - 1 }, false)
            }
            _ => {
                let k = self.rightmost_eq(x, y, b);
                if k != 0 {
                    return (b, false); // weak expansion
                }
                let j = self.rightmost_eq(b, x, b); // strong expansion
                (if j != 0 { j } else { b }, true) // not found -> as weak
            }
        }
    }

    /// Copy rule; tPPS4/fPPS4 strong-branch modifiers on the last-term chain.
    fn copied(&self, src: i64, x: i64, l: i64, y: i64, i: i64, strong: bool) -> i64 {
        let base = if src >= x { src + l } else { src };
        if !strong {
            return base;
        }
        if matches!(self.variant, PpsVariant::TPps4 | PpsVariant::FPps4) {
            if i % l == y % l {
                if self.variant == PpsVariant::FPps4 && i == y {
                    return x;
                }
                return src + l;
            }
        }
        base
    }

    /// expand_len(A, M): replace last + append M elements (source i = q + y - L).
    fn expand_len(&mut self, m: i64) -> Result<(), String> {
        if self.seq.is_empty() {
            return Ok(());
        }
        let y = self.seq.len() as i64;
        let x = *self.seq.last().unwrap();
        if x == 0 {
            self.seq.pop();
            return Ok(());
        }
        if x >= y {
            return Err("Pps: bad root out of range (x >= y)".to_string());
        }
        let b = self.seq[(x - 1) as usize];
        let l = y - x;
        let (new_last, strong) = self.replace_last(x, b, y);
        self.seq[(y - 1) as usize] = new_last;
        for q in 1..=m {
            let i = q + y - l;
            if i < 1 || (i as usize) > self.seq.len() {
                return Ok(()); // defensive
            }
            let src = self.seq[(i - 1) as usize];
            self.seq.push(self.copied(src, x, l, y, i, strong));
        }
        Ok(())
    }
}

impl Notation for Pps {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn name(&self) -> &str {
        self.variant.reg_name()
    }

    fn family(&self) -> Family {
        Family::Ordinal
    }

    fn subfamily(&self) -> &str {
        "sequence"
    }

    fn style(&self) -> &str {
        "marked_parent"
    }

    fn creator(&self) -> &str {
        "318`4"
    }

    fn version(&self) -> &str {
        self.variant.ver()
    }

    fn capabilities(&self) -> Capabilities {
        let mut c = Capabilities::new();
        c.set(Op::FromString);
        c.set(Op::ToString);
        c.set(Op::Normalize);
        c.set(Op::Compare);
        c.set(Op::Expand);
        c.set(Op::ExpandTo);
        c.set(Op::Successor);
        c
    }

    fn string_to_it(&mut self, s: &str) -> Result<(), String> {
        self.seq.clear();
        self.is_master_limit = false;
        let t: String = s.chars().filter(|c| !c.is_whitespace()).collect();
        let t = t
            .trim_start_matches(['(', '[', '{'])
            .trim_end_matches([')', ']', '}']);
        if t.is_empty() {
            return Ok(());
        }
        for p in t.split(',') {
            if !p.is_empty() {
                self.seq
                    .push(p.parse::<i64>().map_err(|e| format!("Pps: {}", e))?);
            }
        }
        Ok(())
    }

    fn to_string(&self) -> String {
        self.to_latex()
    }

    /// Library 0-index: expand(k) = 基本列[k+1] = expand_len((k+1)L - 1).
    fn expand(&mut self, n: i64) -> Result<(), String> {
        if self.is_master_limit {
            self.seq = (0..n).collect();
            self.is_master_limit = false;
            return Ok(());
        }
        if self.seq.is_empty() {
            return Ok(());
        }
        let x = *self.seq.last().unwrap();
        if x == 0 {
            self.seq.pop();
            return Ok(());
        }
        let y = self.seq.len() as i64;
        if x >= y {
            return Err("Pps: bad root out of range (x >= y)".to_string());
        }
        let l = y - x;
        self.expand_len((n + 1) * l - 1)
    }

    fn expand_to(&mut self, len: i64) -> Result<(), String> {
        self.expand_len(len)
    }

    fn compare(&self, other: &dyn Notation) -> Result<i32, String> {
        let o = other
            .as_any()
            .downcast_ref::<Pps>()
            .ok_or_else(|| NotComparable { name: self.name().to_string() }.to_string())?;
        if self.variant != o.variant {
            return Err(NotComparable { name: self.name().to_string() }.to_string());
        }
        if self.is_master_limit || o.is_master_limit {
            return Ok(match (self.is_master_limit, o.is_master_limit) {
                (true, true) => 0,
                (true, false) => 1,
                (false, true) => -1,
                (false, false) => 0,
            });
        }
        let m = self.seq.len().min(o.seq.len());
        for i in 0..m {
            if self.seq[i] > o.seq[i] {
                return Ok(1);
            }
            if self.seq[i] < o.seq[i] {
                return Ok(-1);
            }
        }
        Ok(match self.seq.len().cmp(&o.seq.len()) {
            std::cmp::Ordering::Greater => 1,
            std::cmp::Ordering::Less => -1,
            std::cmp::Ordering::Equal => 0,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn e(v: PpsVariant, s: &str) -> Pps {
        Pps::from_string(v, s).unwrap()
    }

    #[test]
    fn pps1_parse_and_expand() {
        assert_eq!(e(PpsVariant::Pps1, "(0,1,2)").to_string(), "(0, 1, 2)");
        let mut p = e(PpsVariant::Pps1, "(0,1,2)");
        p.expand(0).unwrap();
        assert_eq!(p.to_string(), "(0, 1, 1)");
        let mut p = e(PpsVariant::Pps1, "(0,1,2)");
        p.expand(2).unwrap();
        assert_eq!(p.to_string(), "(0, 1, 1, 1, 1)");
    }

    #[test]
    fn pps1_no_term_between() {
        let mut p = e(PpsVariant::Pps1, "(0,1,0,3)");
        p.expand(0).unwrap();
        assert_eq!(p.to_string(), "(0, 1, 0, 2)");
    }

    #[test]
    fn pps1_term_equal_b() {
        let mut p = e(PpsVariant::Pps1, "(0,1,0,1,2)");
        p.expand(0).unwrap();
        assert_eq!(p.to_string(), "(0, 1, 0, 1, 1, 0, 1)");
    }

    #[test]
    fn successor() {
        let mut p = e(PpsVariant::Pps1, "(0,1,0)");
        p.expand(3).unwrap();
        assert_eq!(p.to_string(), "(0, 1)");
        assert!(e(PpsVariant::Pps1, "(0,1,0)").is_successor());
        assert!(!e(PpsVariant::Pps1, "(0,1,2)").is_successor());
    }

    #[test]
    fn expand_to() {
        let mut p = e(PpsVariant::Pps1, "(0,1,2)");
        p.expand_to(2).unwrap();
        assert_eq!(p.to_string(), "(0, 1, 1, 1, 1)");
    }

    #[test]
    fn pps1_vs_pps2() {
        let mut p1 = e(PpsVariant::Pps1, "(0,2,4,4,2,3)");
        p1.expand(0).unwrap();
        assert_eq!(p1.to_string(), "(0, 2, 4, 4, 2, 4, 7, 2)");
        let mut p2 = e(PpsVariant::Pps2, "(0,2,4,4,2,3)");
        p2.expand(0).unwrap();
        assert_eq!(p2.to_string(), "(0, 2, 4, 4, 2, 2, 7, 2)");
    }

    #[test]
    fn pps4_weak_strong() {
        let mut p = e(PpsVariant::Pps4, "(0,1,0,2,2,3)");
        p.expand(0).unwrap();
        assert_eq!(p.to_string(), "(0, 1, 0, 2, 2, 1, 2, 2)");
        let mut p = e(PpsVariant::Pps4, "(0,1,2,1,3)");
        p.expand(0).unwrap();
        assert_eq!(p.to_string(), "(0, 1, 2, 1, 2, 1)");
    }

    #[test]
    fn tpps4_fpps4() {
        let mut p = e(PpsVariant::TPps4, "(0,1,0,2,2,3)");
        p.expand(1).unwrap();
        assert_eq!(p.to_string(), "(0, 1, 0, 2, 2, 1, 2, 2, 4, 2, 2)");
        let mut p = e(PpsVariant::FPps4, "(0,1,0,2,2,3)");
        p.expand(1).unwrap();
        assert_eq!(p.to_string(), "(0, 1, 0, 2, 2, 1, 2, 2, 3, 2, 2)");
    }

    #[test]
    fn limit_and_master() {
        assert_eq!(Pps::limit(PpsVariant::Pps1, 3).to_string(), "(0, 1, 2)");
        let mut m = Pps::master_limit(PpsVariant::Pps1);
        m.expand(3).unwrap();
        assert_eq!(m.to_string(), "(0, 1, 2)");
    }

    #[test]
    fn compare_lexicographic() {
        let a = e(PpsVariant::Pps1, "(0,1,2)");
        let b = e(PpsVariant::Pps1, "(0,1,3)");
        assert_eq!(a.compare(&b).unwrap(), -1);
        assert_eq!(b.compare(&a).unwrap(), 1);
        assert_eq!(a.compare(&a).unwrap(), 0);
        let c = e(PpsVariant::Pps1, "(0,1,2,0)");
        assert_eq!(a.compare(&c).unwrap(), -1);
    }

    #[test]
    fn cross_variant_not_comparable() {
        let a = e(PpsVariant::Pps1, "(0,1,2)");
        let b = e(PpsVariant::Pps2, "(0,1,2)");
        assert!(a.compare(&b).is_err());
    }

    #[test]
    fn taxonomy() {
        assert_eq!(e(PpsVariant::Pps1, "(0,1,2)").name(), "pps1");
        assert_eq!(e(PpsVariant::TPps4, "(0,1,2)").name(), "tpps4");
        assert_eq!(e(PpsVariant::Pps1, "(0,1,2)").style(), "marked_parent");
        assert_eq!(e(PpsVariant::Pps1, "(0,1,2)").creator(), "318`4");
        assert_eq!(e(PpsVariant::TPps4, "(0,1,2)").version(), "4t");
    }
}
